### PR TITLE
Add argon/classic osu!mania combo counter

### LINKED ISSUE #20541: Mania LegacyComboCounter get a shadow backdrop when counter increases
### Type

Cosmetic

### Bug description

LegacyComboCounter  gets a shadow behind itself when the counter increases like the stable counter. This is very distracing when using the combo counter as a reference point. 

Skin: https://mega.nz/file/C2QVjKZC#OxGzqLXNNp7rtmc51E-xMYtdwpcZ7LAQ8m2XzDAwteI

also tested with this skin: https://osu.ppy.sh/community/forums/topics/203238?n=1

### Screenshots or videos

Lazer behaviour:
Screenshot 2022-10-01 211429

Stable behaviour:
Screenshot 2022-10-01 211714

### Version

2022.923.1

### Logs

[updater.log](https://github.com/ppy/osu/files/9691048/updater.log)
[legacy-ipc.log](https://github.com/ppy/osu/files/9691050/legacy-ipc.log)
[performance.log](https://github.com/ppy/osu/files/9691053/performance.log)
[runtime.log](https://github.com/ppy/osu/files/9691054/runtime.log)
[updater.log](https://github.com/ppy/osu/files/9691055/updater.log)
[updater.log](https://github.com/ppy/osu/files/9691058/updater.log)

### PR DESCRIPTION
**Pull request by frenzibyte**
- Closes https://github.com/ppy/osu/issues/20541
- Supersedes #20579 
- [x] Depends on https://github.com/ppy/osu/pull/26249

Adds support for classic osu!mania combo counter, as well as a very basic implementation for "Argon" skin to increase its usability (cc @peppy for adding animations to the argon one if wanted).

This does not include turning the combo colour to yellow during a hold note, especially since hold notes don't have ticks anymore, which makes the yellow state look odd.

Preview:

https://github.com/ppy/osu/assets/22781491/fae36613-f0e8-4bed-8c1c-81ee7610919f

Argon preview (again, not final):

https://github.com/ppy/osu/assets/22781491/2ad44d09-3ec2-4714-8d73-ff2f345b2ce7

### CODE REVIEW

**Comment by mcendu** [comment]
I think this can be used for taiko as well?

**Comment by frenzibyte** [comment]
Yes but it's not in scope of this PR.

File: osu.Game/Screens/Play/Player.cs
Comment (peppy) [inline osu.Game/Screens/Play/Player.cs:240]: This is a bit ad-hoc isn't it

File: osu.Game/Skinning/ISerialisableDrawable.cs
Comment (peppy) [inline osu.Game/Skinning/ISerialisableDrawable.cs:34]: What's an example of this? I don't quite understand.

File: osu.Game/Skinning/LegacyDefaultComboCounter.cs
Comment (peppy) [inline osu.Game/Skinning/LegacyDefaultComboCounter.cs:13]: I'm honestly not sure this abstraction was required. Could have just copy-pasted and spent less effort? And potentially had more maintainable / legible code.

File: osu.Game/Screens/Play/Player.cs
Comment (frenzibyte) [reply osu.Game/Screens/Play/Player.cs:240]: I...probably can just resolve the ruleset and get the scrolling info from there.

File: osu.Game/Skinning/ISerialisableDrawable.cs
Comment (frenzibyte) [reply osu.Game/Skinning/ISerialisableDrawable.cs:34]: Disabled by `LegacyManiaComboCounter` because its position is shifted when the user flips the scrolling direction. I don't recall why though.

File: osu.Game/Skinning/LegacyDefaultComboCounter.cs
Comment (frenzibyte) [reply osu.Game/Skinning/LegacyDefaultComboCounter.cs:13]: Sure.

File: osu.Game/Skinning/ISerialisableDrawable.cs
Comment (peppy) [reply osu.Game/Skinning/ISerialisableDrawable.cs:34]: This seems like an odd choice. So the user can still choose an anchor and it will just flip? I'd say that closest should still be allowed and the flip is only applied if the counter is in a certain region of the screen (or have a certain anchor applied).

File: osu.Game/Screens/Play/Player.cs
Comment (peppy) [reply osu.Game/Screens/Play/Player.cs:240]: Yeah not sure. Maybe it's fine. Just a bit hard to track what DI dependencies are available right now.

File: osu.Game/Skinning/LegacyDefaultComboCounter.cs
Comment (peppy) [reply osu.Game/Skinning/LegacyDefaultComboCounter.cs:13]: No need to change it at this point, I'll do so if the review is painful enough.

File: osu.Game/Screens/Play/Player.cs
Comment (peppy) [reply osu.Game/Screens/Play/Player.cs:240]: I'm gonna leave this as-is for now. Rationale is that `IScrollingInfo` is used in a lot of places within `DrawableRuleset` so I'd expect for it to also be available here.

File: osu.Game/Skinning/ISerialisableDrawable.cs
Comment (peppy) [reply osu.Game/Skinning/ISerialisableDrawable.cs]: I've removed this for now as I can't really understand why it is required. It would make more sense to completely disable anchors (or a certain subset) but honestly I'd just leave this up to the user, and adjust the automatic flipping logic to only work in the cases where it makes sense.

File: osu.Game.Rulesets.Mania/Skinning/Argon/ArgonManiaComboCounter.cs
Comment (peppy) [inline osu.Game.Rulesets.Mania/Skinning/Argon/ArgonManiaComboCounter.cs:27]: really dunno but willing to overlook for now because it works.

**Comment by peppy** [comment]
@frenzibyte I've made sizable changes to this PR. Want to give it a once-over?

**Review by peppy — DISMISSED** [review]

File: osu.Game.Rulesets.Osu/Skinning/Legacy/OsuLegacySkinTransformer.cs
Comment (frenzibyte) [inline osu.Game.Rulesets.Osu/Skinning/Legacy/OsuLegacySkinTransformer.cs:50]: Nit-pick but I don't expect this to be placed here:
```diff
diff --git a/osu.Game.Rulesets.Catch/Skinning/Legacy/CatchLegacySkinTransformer.cs b/osu.Game.Rulesets.Catch/Skinning/Legacy/CatchLegacySkinTransformer.cs
index df8c04638d..2ba87eda2c 100644
--- a/osu.Game.Rulesets.Catch/Skinning/Legacy/CatchLegacySkinTransformer.cs
+++ b/osu.Game.Rulesets.Catch/Skinning/Legacy/CatchLegacySkinTransformer.cs
@@ -31,9 +31,6 @@ public CatchLegacySkinTransformer(ISkin skin)
             switch (lookup)
             {
                 case SkinComponentsContainerLookup containerLookup:
-                    if (containerLookup.Target != SkinComponentsContainerLookup.TargetArea.MainHUDComponents)
-                        return base.GetDrawableComponent(lookup);
-
                     // Modifications for global components.
                     if (containerLookup.Ruleset == null)
                         return base.GetDrawableComponent(lookup);
@@ -43,27 +40,33 @@ public CatchLegacySkinTransformer(ISkin skin)
                         return d;
 
                     // Our own ruleset components default.
-                    // todo: remove CatchSkinComponents.CatchComboCounter and refactor LegacyCatchComboCounter to be added here instead.
-                    return new DefaultSkinComponentsContainer(container =>
+                    switch (containerLookup.Target)
                     {
-                        var keyCounter = container.OfType<LegacyKeyCounterDisplay>().FirstOrDefault();
-
-                        if (keyCounter != null)
-                        {
-                            // set the anchor to top right so that it won't squash to the return button to the top
-                            keyCounter.Anchor = Anchor.CentreRight;
-                            keyCounter.Origin = Anchor.CentreRight;
-                            keyCounter.X = 0;
-                            // 340px is the default height inherit from stable
-                            keyCounter.Y = container.ToLocalSpace(new Vector2(0, container.ScreenSpaceDrawQuad.Centre.Y - 340f)).Y;
-                        }
-                    })
-                    {
-                        Children = new Drawable[]
-                        {
-                            new LegacyKeyCounterDisplay(),
-                        }
-                    };
+                        case SkinComponentsContainerLookup.TargetArea.MainHUDComponents:
+                            // todo: remove CatchSkinComponents.CatchComboCounter and refactor LegacyCatchComboCounter to be added here instead.
+                            return new DefaultSkinComponentsContainer(container =>
+                            {
+                                var keyCounter = container.OfType<LegacyKeyCounterDisplay>().FirstOrDefault();
+
+                                if (keyCounter != null)
+                                {
+                                    // set the anchor to top right so that it won't squash to the return button to the top
+                                    keyCounter.Anchor = Anchor.CentreRight;
+                                    keyCounter.Origin = Anchor.CentreRight;
+                                    keyCounter.X = 0;
+                                    // 340px is the default height inherit from stable
+                                    keyCounter.Y = container.ToLocalSpace(new Vector2(0, container.ScreenSpaceDrawQuad.Centre.Y - 340f)).Y;
+                                }
+                            })
+                            {
+                                Children = new Drawable[]
+                                {
+                                    new LegacyKeyCounterDisplay(),
+                                }
+                            };
+                    }
+
+                    return null;
 
                 case CatchSkinComponentLookup catchSkinComponent:
                     switch (catchSkinComponent.Component)
```

I also don't want to imagine that there is any other class that will return ruleset-specific components for the legacy skin other than this one, so I'm returning null at the end of the diff above.

Maybe this is irrelevant anyway because these methods will be refactored later on.

**Review by frenzibyte — COMMENTED** [review]
Combo counter does not seem to show up on argon skin:
CleanShot 2024-08-10 at 11 20 20

I also kinda don't like the "x" symbol in the argon counter, looks weird when it's placed on top of the playfield. Would try to remove it.

It's also incorrectly pushed away on classic skin:
CleanShot 2024-08-10 at 11 19 44

File: osu.Game.Rulesets.Osu/Skinning/Legacy/OsuLegacySkinTransformer.cs
Comment (frenzibyte) [reply osu.Game.Rulesets.Osu/Skinning/Legacy/OsuLegacySkinTransformer.cs:50]: Also why not put the "Skin has configuration" if statement to the top of the method?

**Comment by peppy** [comment]
> Combo counter does not seem to show up on argon skin

can you look into this? it shows fine here.

File: osu.Game.Rulesets.Osu/Skinning/Legacy/OsuLegacySkinTransformer.cs
Comment (peppy) [reply osu.Game.Rulesets.Osu/Skinning/Legacy/OsuLegacySkinTransformer.cs]: I've applied something similar to your diff. Still waiting on feedback from you for the issues you encountered during testing.

**Comment by frenzibyte** [comment]
I began investigating this and posted some [observations in discord](https://discord.com/channels/188630481301012481/188630652340404224/1272481231245807711) until all of a sudden everything started working again and I have no idea why.

File: osu.Game.Rulesets.Mania/Skinning/Argon/ArgonManiaComboCounter.cs
Comment (frenzibyte) [inline osu.Game.Rulesets.Mania/Skinning/Argon/ArgonManiaComboCounter.cs:34]: The legacy part handles the case where the anchor is set to center, but this one doesn't?

File: osu.Game.Rulesets.Mania/Skinning/Argon/ArgonManiaComboCounter.cs
Comment (frenzibyte) [inline osu.Game.Rulesets.Mania/Skinning/Argon/ArgonManiaComboCounter.cs:12]: As I mentioned already I think the "x" symbol next to the counter feels weird on mania.

File: osu.Game.Rulesets.Mania/Skinning/Argon/ManiaArgonSkinTransformer.cs
Comment (frenzibyte) [inline osu.Game.Rulesets.Mania/Skinning/Argon/ManiaArgonSkinTransformer.cs:41]: Both this and `ManiaLegacySkinTransformer` haven't been updated in line with d2eb6ccb8caa

**Comment by frenzibyte** [comment]
As you touched in [discord](https://discord.com/channels/188630481301012481/188630652340404224/1272489428669169725), it turns out that the cause of this issue is that I had "beatmap skins" enabled, which makes the beatmap skin layer return a `LegacyComboCounter` with a zero combo position.

The reason why the combo position is zero is because of the lookup only happening on the beatmap skin layer, and the beatmap skin layer disables mania lookups entirely (see `AllowManiaConfigLookups` override), therefore the fallback value in the lookup line (which is `0`) is assigned and therefore the combo counter is stuck at the top.

I'm not sure what is the best way to solve this, one way could be to use the `isLegacySkin` flag which determines whether the skin is user or beatmap and then return null when returning the default mania HUD layout if it's a beatmap skin, but I'm not really sure that's a good way to handle things.

File: osu.Game.Rulesets.Mania/Skinning/Argon/ArgonManiaComboCounter.cs
Comment (peppy) [reply osu.Game.Rulesets.Mania/Skinning/Argon/ArgonManiaComboCounter.cs:34]: I didn't change this 🙈

Comment (frenzibyte) [reply osu.Game.Rulesets.Mania/Skinning/Argon/ArgonManiaComboCounter.cs:34]: Whoops 😅

Comment (peppy) [reply osu.Game.Rulesets.Mania/Skinning/Argon/ArgonManiaComboCounter.cs]: I've updated this to match but there still seems to be something weird going on here.

- Set scroll direction to "Down"
- Put the combo counter near the bottom of the screen
- Change scroll direction

The combo won't move the first time changing direction.

File: osu.Game.Rulesets.Mania/Skinning/Argon/ManiaArgonSkinTransformer.cs
Comment (peppy) [reply osu.Game.Rulesets.Mania/Skinning/Argon/ManiaArgonSkinTransformer.cs:41]: Have updated these.

File: osu.Game.Rulesets.Mania/Skinning/Argon/ArgonManiaComboCounter.cs
Comment (frenzibyte) [reply osu.Game.Rulesets.Mania/Skinning/Argon/ArgonManiaComboCounter.cs]: Alleviated the issue in https://github.com/ppy/osu/pull/26254/commits/66adddbfb8020c019bbdf672ade3f00b4aead7ef. This is the reason why "Closest" anchor should not be used.

To explain briefly, I am using the sign of the Y coordinate as an indicator for whether the combo counter is in the correct position or not. When the scroll direction is "Down", the anchor is set to top and the Y position is made sure to be positive. When the scroll direction is "Up", the anchor is set to bottom and the Y position is made sure to be negative.

When the "Closest" anchor is selected, if the user moves their combo counter from top to bottom or vice versa, the anchor will change, and defies everything that the logic above does, therefore breaking apart.

**Review by peppy — DISMISSED** [review]

**Pull request merged**
